pydantic/main.py

```diff
@@ -56,6 +58,42 @@
 
 _object_setattr = _model_construction.object_setattr
 
+_K = TypeVar('_K')
+_V = TypeVar('_V')
+
+
+# We need a sentinel value for missing fields when comparing models
+# Models are equals if-and-only-if they miss the same fields, and since None is a legitimate value
+# we can't default to None
+# We use the single-value enum trick to allow correct typing when using a sentinel
+class _SentinelType(enum.Enum):
+    SENTINEL = enum.auto()
+
+
+_SENTINEL = _SentinelType.SENTINEL
+
+
+class _SafeGetItemProxy(Generic[_K, _V]):
```

Comment:
We should think about either combining our PRs now or have an idea of how we might clean up duplicate code after they both merge. This class is essentially duplicated in https://github.com/pydantic/pydantic/pull/7786/files#diff-ea75a1ef1da6bb38913477e979605f6a05e2ffe8c28fe47d9b798c89ba436c7cR422-R427 so we should agree on what it should look like. IMHO the type annotations are overkill for something so simple. I also think the `__contains__` and `__repr__` should probably be removed. `__contains__` in particular just seems to be a quirk of the type checker and gives the reader the false impression that it's actually being used. If you think they should be kept, add `# pragma: no cover` to avoid reducing test coverage.